cstar/base/component.py

```diff
@@ -37,7 +40,13 @@ class Component(ABC):
         Execute any post-processing actions associated with this component
     """
 
-    def __init__(self, **kwargs: Any):
+    def __init__(
+        self,
+        base_model: BaseModel,
+        additional_code: Optional["AdditionalCode"],
+        input_datasets: List[InputDataset] = [],
```

Comment:
Technically mutable default arguments should be [avoided](https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil).

Instead this should be

```python
def __init__(
    ...
    input_datasets: List[InputDataset] | None = None,
    ...
):
    if input_datasets is None:
        input_datasets = []
```

I'm actually surprised the linter didn't point this out.

Reply:
This was actually to silence the linter - it wouldn't stop complain that None is not iterable, even when it's explicitly set to a list immediately. I think the above suggestion is implicit optional which will also wind the linter up.
This should work:
```
def __init__(
    ...
    input_datasets: Optional[List[InputDataset]]
    ...
):
   if input_datasets is None:
      input_datasets = []
```
... but if I remember correctly whenever mypy saw Optional it required an explicit check for `None` before doing anything listy with `input_datasets`, even though it's never `None`. The issue with the mutable default is dealt with a few lines later:

`self.input_datasets: List[InputDataset] = list(input_datasets)`

I can't remember the specifics but this was a nightmare and this outcome took a good while to land on as a solution. Would you be up for having a crack at it?

Reply:
> This was actually to silence the linter - it wouldn't stop complain that None is not iterable, even when it's explicitly set to a list immediately.

That's annoying, I don't know why mypy doesn't just narrow down what the type must be.

I think when I've had this problem before the simplest solution was just to do

```python
def __init__(
    ...
    input_datasets: Optional[List[InputDataset]]
    ...
):
   if input_datasets is None:
      _input_datasets = []
```

Reply:
or on one line:
```python
_input_datasets = [] if input_datasets is None else input_datasets
```

Reply:
Reading into this, I'm pretty sure my suggestion is the least worst way. Unfortunately I think this is one of those ugly mypy things that is a consequence of static typing being added to python post-hoc, not when the language was invented.

Reply:
I'll give it a go, but I'm not sure this will address the underlying issue: IIRC the rest of the method was the only place where mypy _was_ happy with the typing - it was everywhere else that `Component.input_datasets` got used throughout the codebase... I'd have to revisit this again to be sure though.

Reply:
... I think it was what I've committed here vs. `assert isinstance(Component.input_datasets, list)` in every other method that used it...

Reply:
`Component.input_datasets` is a different variable to the `input_datasets` being used inside this init method. That sounds like a different issue. You can specify what the type of `Component.input_datasets` should be like this:

```python
class Component:
    input_datasets: list[InputDataset]

    def __init__(self, ...):
        ...
```

> ... I think it was what I've committed here vs. `assert isinstance(Component.input_datasets, list)` in every other method that used it...

Though I'm not totally sure I understand what you're saying here.

Reply:
Right, I might even be confusing memories of mypy's various whinings here, so I think I'm too out of the loop to say anything about this with confidence without revisiting it for now.

Reply:
Your line below

```python
self.input_datasets: List[InputDataset] = list(input_datasets)
```
should achieve the same thing though... But if you type hint it in the class definition you don't need to type hint it during assignment.

Reply:
I actually meant to ask you to keep a general eye out for this and was going to raise an issue to broadly revisit tidy up type hints - I think I've over-hinted throughout and used `Optional` in a way that probably isn't needed but makes mypy much stricter...

Reply:
Let me have a go at cleaning this up and pushing to this branch now

Reply:
Should be fixed in https://github.com/CWorthy-ocean/C-Star/pull/71/commits/c2c311cf751d2a3a9f2019a5c5fd1444e9823856